File: tiktok-studio-hub/web/lib/pkce.ts
import crypto from 'crypto'

const VERIFIER_CHARS =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~'

export function generateCodeVerifier (length = 64): string {
  let result = ''
  for (let i = 0; i < length; i++) {
    result += VERIFIER_CHARS[crypto.randomInt(0, VERIFIER_CHARS.length)]
  }
  return result
}

/** TikTok Desktop PKCE : HEX(SHA256(verifier)), pas base64url. */
export function generateCodeChallenge (codeVerifier: string): string {
  return crypto.createHash('sha256').update(codeVerifier, 'utf8').digest('hex')
}

export function newOAuthState (): { state: string; codeVerifier: string; codeChallenge: string } {
  const state = crypto.randomBytes(24).toString('base64url')
  const codeVerifier = generateCodeVerifier()
  const codeChallenge = generateCodeChallenge(codeVerifier)
  return { state, codeVerifier, codeChallenge }
}

File: tiktok-studio-hub/app/pkce.py
import hashlib
import secrets
import string

_VERIFIER_CHARS = string.ascii_letters + string.digits + "-._~"


def generate_code_verifier(length: int = 64) -> str:
    if length < 43 or length > 128:
        raise ValueError("code_verifier must be between 43 and 128 characters")
    return "".join(secrets.choice(_VERIFIER_CHARS) for _ in range(length))


def generate_code_challenge(code_verifier: str) -> str:
    """TikTok Desktop PKCE : HEX(SHA256(verifier)), pas base64url."""
    digest = hashlib.sha256(code_verifier.encode("utf-8")).digest()
    return digest.hex()
